From 6806bc0f5df04c2f67ac791b6b90a0e1c9f66010 Mon Sep 17 00:00:00 2001
From: Mikulas Patocka <mpatocka@redhat.com>
Date: Tue, 28 May 2024 13:32:34 +0200
Subject: [PATCH 1/2] BACKPORT: FROMGIT: dm: optimize flushes

Backport the minimum dm-core support needed for targets that can bypass
their ->map path for flushes.

This is the support layer required for Android/common
`dm-default-key: Set .flush_bypasses_map` to have real effect on the
Android 14 / Linux 6.1 baseline used here.
---
 drivers/md/dm-core.h          |  2 ++
 drivers/md/dm-table.c         |  4 ++++
 drivers/md/dm.c               | 24 +++++++++++++++++++++++-
 include/linux/device-mapper.h | 14 ++++++++++++++
 4 files changed, 43 insertions(+), 1 deletion(-)

diff --git a/drivers/md/dm-core.h b/drivers/md/dm-core.h
index 371d262..37d2f7f 100644
--- a/drivers/md/dm-core.h
+++ b/drivers/md/dm-core.h
@@ -205,6 +205,8 @@ struct dm_table {
 
 	bool integrity_supported:1;
 	bool singleton:1;
+	/* Set if all targets in the table have "flush_bypasses_map" set. */
+	bool flush_bypasses_map:1;
 	unsigned integrity_added:1;
 
 	/*
diff --git a/drivers/md/dm-table.c b/drivers/md/dm-table.c
index a67f9d2..d16c78e 100644
--- a/drivers/md/dm-table.c
+++ b/drivers/md/dm-table.c
@@ -157,6 +157,7 @@ int dm_table_create(struct dm_table **result, fmode_t mode,
 
 	t->type = DM_TYPE_NONE;
 	t->mode = mode;
+	t->flush_bypasses_map = true;
 	t->md = md;
 	*result = t;
 	return 0;
@@ -723,6 +724,9 @@ int dm_table_add_target(struct dm_table *t, const char *type,
 	if (ti->limit_swap_bios && !static_key_enabled(&swap_bios_enabled.key))
 		static_branch_enable(&swap_bios_enabled);
 
+	if (!ti->flush_bypasses_map)
+		t->flush_bypasses_map = false;
+
 	return 0;
 
  bad:
diff --git a/drivers/md/dm.c b/drivers/md/dm.c
index f70129b..d3b5f96 100644
--- a/drivers/md/dm.c
+++ b/drivers/md/dm.c
@@ -635,7 +635,7 @@ static struct bio *alloc_tio(struct clone_info *ci, struct dm_target *ti,
 
 	/* Set default bdev, but target must bio_set_dev() before issuing IO */
 	clone->bi_bdev = md->disk->part0;
-	if (unlikely(ti->needs_bio_set_dev))
+	if (ti && unlikely(ti->needs_bio_set_dev))
 		bio_set_dev(clone, md->disk->part0);
 
 	if (len) {
@@ -1090,7 +1090,7 @@ static void clone_endio(struct bio *bio)
 	blk_status_t error = bio->bi_status;
 	struct dm_target_io *tio = clone_to_tio(bio);
 	struct dm_target *ti = tio->ti;
-	dm_endio_fn endio = ti->type->end_io;
+	dm_endio_fn endio = ti ? ti->type->end_io : NULL;
 	struct dm_io *io = tio->io;
 	struct mapped_device *md = io->md;
 
@@ -1136,7 +1136,7 @@ static void clone_endio(struct bio *bio)
 	}
 
 	if (static_branch_unlikely(&swap_bios_enabled) &&
-	    unlikely(swap_bios_limit(ti, bio)))
+	    ti && unlikely(swap_bios_limit(ti, bio)))
 		up(&md->swap_bios_semaphore);
 
 	free_tio(bio);
@@ -1536,13 +1536,29 @@ static void __send_empty_flush(struct clone_info *ci)
 	ci->sector_count = 0;
 	ci->io->tio.clone.bi_iter.bi_size = 0;
 
-	for (unsigned int i = 0; i < t->num_targets; i++) {
-		unsigned int bios;
-		struct dm_target *ti = dm_table_get_target(t, i);
+	if (!t->flush_bypasses_map) {
+		for (unsigned int i = 0; i < t->num_targets; i++) {
+			unsigned int bios;
+			struct dm_target *ti = dm_table_get_target(t, i);
 
-		atomic_add(ti->num_flush_bios, &ci->io->io_count);
-		bios = __send_duplicate_bios(ci, ti, ti->num_flush_bios, NULL);
-		atomic_sub(ti->num_flush_bios - bios, &ci->io->io_count);
+			atomic_add(ti->num_flush_bios, &ci->io->io_count);
+			bios = __send_duplicate_bios(ci, ti, ti->num_flush_bios, NULL);
+			atomic_sub(ti->num_flush_bios - bios, &ci->io->io_count);
+		}
+	} else {
+		struct list_head *devices = dm_table_get_devices(t);
+		struct dm_dev_internal *dd;
+		unsigned int len = 0;
+
+		list_for_each_entry(dd, devices, list) {
+			struct bio *clone;
+
+			clone = alloc_tio(ci, NULL, 0, &len, GFP_NOIO);
+			atomic_add(1, &ci->io->io_count);
+			bio_set_dev(clone, dd->dm_dev->bdev);
+			clone->bi_end_io = clone_endio;
+			dm_submit_bio_remap(clone, NULL);
+		}
 	}
 
 	/*
diff --git a/include/linux/device-mapper.h b/include/linux/device-mapper.h
index 13f4936..a8e3b2c 100644
--- a/include/linux/device-mapper.h
+++ b/include/linux/device-mapper.h
@@ -384,6 +384,20 @@ struct dm_target {
 	 */
 	bool needs_bio_set_dev:1;
 
+	/*
+	 * Set if the target supports flush optimization. If all the targets in
+	 * a table have flush_bypasses_map set, the dm core will not send
+	 * flushes to the targets via a ->map method. It will iterate over
+	 * dm_table->devices and send flushes to the devices directly. This
+	 * optimization reduces the number of flushes being sent when multiple
+	 * targets in a table use the same underlying device.
+	 *
+	 * This optimization may be enabled on targets that just pass the
+	 * flushes to the underlying devices without performing any other
+	 * actions on the flush request.
+	 */
+	bool flush_bypasses_map:1;
+
 	ANDROID_KABI_RESERVE(1);
 	ANDROID_KABI_RESERVE(2);
 };
-- 
2.49.0

